java/arcs/core/host/ParticleStateMachine.kt

```diff
@@ -0,0 +1,172 @@
+    init {
+        declareStateMachine {
+            state(Instantiated) {
+                onEntry {
+                    particle.onCreate()
+                    trigger(CreateEvent)
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(StartEvent, Created)
+                on(FailedEvent, Failed_NeverStarted)
+            }
+
+            state(Started) {
+                // Starting while already started will stop, this occurs during crash recovery
+                on(StartEvent, Stopped) {
+                    // Trigger a restart after moving to Stopped state
+                    trigger(StartEvent)
+                }
+                on(StopEvent, Stopped)
+            }
+
+            state(Stopped) {
+                onEntry {
+                    try {
+                        particle.onShutdown()
+                    } finally {
+                        particle.handles.reset()
+                    }
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(StartEvent, Created)
+                on(FailedEvent, Failed)
+            }
+
+            state(Created) {
+                onEntry {
+                    val waitingSet = mutableSetOf<Handle>()
+                    waitingSet.addAll(handles.values)
+
+                    handles.values.forEach { handle ->
+                        handle.onReady {
+                            waitingSet.remove(handle)
+                            if (waitingSet.isEmpty()) {
+                                GlobalScope.launch {
+                                    triggerImmediate(ReadyEvent)
+                                }
+                            }
+                        }
+                    }
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(ReadyEvent, Ready)
+                on(FailedEvent, Failed)
+            }
+
+            state(Ready) {
+                onEntry {
+                    particle.onReady()
+                    trigger(StartEvent)
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(FailedEvent, Failed)
+                on(StartEvent, Started)
+            }
+
+            state(Failed) {
+                onEntry(checkForMaxFailures)
+                onError {
+                    trigger(MaxFailedEvent)
+                }
+                on(MaxFailedEvent, MaxFailed)
+                on(StartEvent, Created)
+                on(StopEvent, Failed)
+            }
+
+            state(Failed_NeverStarted) {
+                onEntry(checkForMaxFailures)
+                onError {
+                    trigger(MaxFailedEvent)
+                }
+                on(StartEvent, Instantiated)
+                on(MaxFailedEvent, MaxFailed)
+                on(StopEvent, Failed_NeverStarted)
+            }
+
+            state(MaxFailed) {
+                // There is no escape
+            }
+        }
+    }
```

Comment:
This is a lot to take-in. It would be good to have nice comment blocks for each state explaining what the state means, and what its transitions imply.